### ISSUE TITLE
Common_oM: Material

### ISSUE THREAD

**Issue by ZiolkowskiJakub** [BHoM/BHoM#393]
Hi,

I would like start here s discussion about Materials and how we can make common approach for Material definition.

Currently there are separate definitions for Environmental and Structural material.

We should discuss where Material definition goes. Is it Discipline specific? If yes then why structural material is in Common namespace?

Is it possible to implement core Interface or all Materials? As far as I can see there are some common properties.

**Comment by rwemay** [comment]
Hi @ZiolkowskiJakub , material definitions should be in the common namespace, and definitely discipline agnostic.

**Comment by michaldengusiak** [comment]
I would think we have some:
-  first level default basic properties common among all discipline 
- second level discipline specific we need to think how we divide and store data need to daylight (more basic) and energy simulation 
- and last third level  software specific variant for above 

I guess would be good to Sketch some example

1. Name= Concrete Color:  hex value definition
2. **Daylight**: Refelctance Internal,External;  Transmission Internal,External   **EnergySimulation**: Conductivity, SpecificHeat, Density, VapourDiffusionFactor, SolarReflectanceExternal, SolarReflectanceInternal, LightReflectanceExternal, LightReflectanceInternal, EmissivityInternal, EmissivityExternal, Roughness    

3. Software Specific **TAS**: IgnoreInUvalue, IsBlind or **Radiance**:

![image](https://user-images.githubusercontent.com/26113670/51303857-842efb00-1a2e-11e9-87ab-e1ea715353fa.png)

so 

RGB inoput could be recalculated form Color:  hex value definition [see example here](https://stackoverflow.com/questions/12155811/how-to-convert-hexadecimal-ffffff-to-system-drawing-color)

```
## ----------------------- Concrete ----------------------- ##
#
# concrete / (Color_45058_46543_46996)
# 
dusty_wide plastic (Color_45058_46543_46996)
0
0
5 0.687541 0.710201 0.717113 0 0
```

**Comment by IsakNaslundBh** [comment]
@rwemay Might be worth considering having all the base and common parameters on the base material class, such as density, and then have like different modules on this class covering the different needs, such as structural, env, acoustic etc?

Given that we will need to expand our structural material capabilities to for example orthotropic materials, and also probably make a divide between the isotropic materials for design properties, it feels like it can get a bit out of hand to have it all on the base class.

Assume the case is the same for env.

Basically what @michaldengusiak described above.

**Comment by rwemay** [comment]
Yep, sounds perfect. A StructuralProperties (or similarly named/removing 'properties') class on the Common material class, from the structure oM then, I presume? 

@michaldengusiak , good point on software-specific and needing a home for that. Referring to the conversation on CustomData etc, though, I think we need a more global solution to carrying software-specific properties on objects, that's inherited from the BHoM base class, rather than adding lots of software-specific classes to every object individually.

**Comment by IsakNaslundBh** [comment]
@rwemay have to be a bit vary about the dependencies here. Currently Structure_oM has a dependency on Common_oM, thanks to the material class.

This ofc means that to opposite is impossible. What we could do is to have a `IStructuralMaterialProperties`, `IEnvironmentalMateralProperties` etc interfaces in the Common_oM, that the discipline parts then implement.

Sort of like:

```
    public class Material
    {
        //Some common properties like density

        public static IStructuralMaterialProperties {get;set;}
        public static IEnvironmentalMaterialProperties {get;set;}
    }

```

and then

`public class StructuralConcrete : IStructuralMaterialProperties `

or something like that.

**Comment by ZiolkowskiJakub** [comment]
What about:

```
    public class Material
    {
        //Some common properties like density

        public static List<IMaterialProperties> {get;set;}
    }
```

and Engine Method

```
public static IMaterialProperties GetMaterialProperites(Type type)
{

}
```

and then

```
public class StructuralConcrete : IMaterialProperties
{

}
```

This would allow to include many sets without changing Material definition

**Comment by al-fisher** [comment]
Yes @ZiolkowskiJakub liking your approach - as discussed ultimately Material is a good candidate for a more formalised Entity Component System (ECS) model similar to your proposal - whereby the Common Material can be easily extended to have additional discipline properties on the fly - even new discipline extensions that we can not foresee today.
This model has already been started to be prototyped in the `human` class by @adecler - with a human having different _Roles_ depending on the behaviours required. So good to review - and perhaps plan similar.
Think good to prototype to work for `Structure` and `Environment` as you are proposing with a  good eye on generality. @adecler perhaps another candidate for us to review in parallel for ECS with further improvement to be planned as appropriate later.

**Comment by FraserGreenroyd** [comment]
@IsakNaslundBh I think this issue can be closed now with the creation of the Physical Material that acts as a basis between our disciplines for materials and provides the common approach to it (albeit not always used)?

**Comment by IsakNaslundBh** [comment]
@FraserGreenroyd yes, it can. The physical material has basically solved this.

Would say outstanding parts for it is getting the fraction system in place and the simpler one of deleting the current material in Common_oM.
